### ISSUE TITLE
cmake troubles

### ISSUE THREAD

**Issue by UNCuser** [3dem/relion#216]
Hello,

I am having a cmake error shown below 
Any insight would be appreciated.

Happy to provide further clarification if necessary.
OS=Red Hat Enterprise Linux Server release 7.2

```
cat CMakeError.log
Determining if the __sincos exist failed with the following output:
Change Dir: /nas/GPUcluster/home/UNCuser/relion2/relion/build2/CMakeFiles/CMakeTmp

Run Build Command:"/usr/bin/gmake" "cmTC_878df/fast"
/usr/bin/gmake -f CMakeFiles/cmTC_878df.dir/build.make CMakeFiles/cmTC_878df.dir/build
gmake[1]: Entering directory `/nas/GPUcluster/home/UNCuser/relion2/relion/build2/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_878df.dir/CheckSymbolExists.cxx.o
/nas/GPUcluster/home/UNCuser/packages/openmpi-2.0.2/bin/mpicxx      -o CMakeFiles/cmTC_878df.dir/CheckSymbolExists.cxx.o -c /nas/GPUcluster/home/UNCuser/relion2
/relion/build2/CMakeFiles/CMakeTmp/CheckSymbolExists.cxx
Linking CXX executable cmTC_878df
/nas/GPUcluster/apps/cmake/3.5.2/bin/cmake -E cmake_link_script CMakeFiles/cmTC_878df.dir/link.txt --verbose=1
/nas/GPUcluster/home/UNCuser/packages/openmpi-2.0.2/bin/mpicxx        CMakeFiles/cmTC_878df.dir/CheckSymbolExists.cxx.o  -o cmTC_878df -rdynamic
CMakeFiles/cmTC_878df.dir/CheckSymbolExists.cxx.o: In function `main':
CheckSymbolExists.cxx:(.text+0x23): undefined reference to `__sincos'
gmake[1]: *** [cmTC_878df] Error 1
gmake[1]: Leaving directory `/nas/GPUcluster/home/UNCuser/relion2/relion/build2/CMakeFiles/CMakeTmp'
gmake: *** [cmTC_878df/fast] Error 2

File /nas/GPUcluster/home/UNCuser/relion2/relion/build2/CMakeFiles/CMakeTmp                   /CheckSymbolExists.cxx:
/* */
#include <math.h>

int main(int argc, char** argv)
{
  (void)argv;
#ifndef __sincos
  return ((int*)(&__sincos))[argc];
#else
  (void)argc;
  return 0;
#endif
}

```

**Comment by bforsbe** [comment]
If you search for 'sincos' in this repo, you'll find that I mention in many issues that this is not really an error, but more a warning that gets logged in the error file. Cmake is just testing the existence and upon finding this error it assess that it cannot use certain functions, that all. 

Is the code not compiling or what is the issue itself?

**Comment by UNCuser** [comment]
Hi thanks for getting back to me.
My apologizes on sincos - title changed.

Perhaps this is more relevant
I am unable to complete cmake and produce a Makefile.

```
CMake Error at CMakeLists.txt:163 (include_directories):
  include_directories given empty-string as include directory
```

Of note, cluster has no mpi support, so open mpi was installed locally.


```
-- The C compiler identification is Intel 16.0.2.20160204
-- The CXX compiler identification is Intel 16.0.2.20160204

**also tried with** 

- The C compiler identification is GNU 4.8.5
-- The CXX compiler identification is GNU 4.8.5

-- Check for working C compiler: /nas/GPUcluster/home/UNCuser/packages/openmpi-2.0.2/bin/mpicc
-- Check for working C compiler: /nas/GPUcluster/home/UNCuser/packages/openmpi-2.0.2/bin/mpicc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /nas/GPUcluster/home/UNCuser/packages/openmpi-2.0.2/bin/mpicxx
-- Check for working CXX compiler: /nas/GPUcluster/home/UNCuser/packages/openmpi-2.0.2/bin/mpicxx -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- BUILD TYPE set to the default type:  'Release'
-- Setting fallback CUDA_ARCH=35
-- Setting cpu precision to double
-- Setting gpu precision to single
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE
-- Found CUDA: /usr/local/cuda (found version "8.0")
-- Using cuda wrapper to compile....
-- Cuda version is >= 7.5 and single-precision build, enable double usage warning.
-- Found MPI_C: /nas/GPUcluster/home/UNCuser/packages/openmpi-2.0.2/bin/mpicc
-- Found MPI_CXX: /nas/GPUcluster/home/UNCuser/packages/openmpi-2.0.2/bin/mpicxx
CMake Error at CMakeLists.txt:163 (include_directories):
  include_directories given empty-string as include directory.


-- MPI_INCLUDE_PATH :
-- MPI_LIBRARIES :
-- MPI_CXX_INCLUDE_PATH :
-- MPI_CXX_LIBRARIES :
-- Looking for XOpenDisplay in /usr/lib64/libX11.so;/usr/lib64/libXext.so
-- Looking for XOpenDisplay in /usr/lib64/libX11.so;/usr/lib64/libXext.so - found
-- Looking for gethostbyname
-- Looking for gethostbyname - found
-- Looking for connect
-- Looking for connect - found
-- Looking for remove
-- Looking for remove - found
-- Looking for shmat
-- Looking for shmat - found
-- Looking for IceConnectionNumber in ICE
-- Looking for IceConnectionNumber in ICE - found
-- Found X11: /usr/lib64/libX11.so
-- Could NOT find FLTK (missing:  FLTK_LIBRARIES FLTK_INCLUDE_DIR FLTK_FLUID_EXECUTABLE)
-- FLTK was NOT found
-- -------------------------------------------------
-- ------- WILL USE LOCALLY BUILT FLTK LIBS --------
-- -------------------------------------------------
-- no previous fltk found, the following paths are set for libs/headers TO BE built
-- FLTK_INCLUDE_DIR: /nas/GPUcluster/home/UNCuser/relion2/relion/external/fltk/include
-- FLTK_LIBRARIES:   /nas/GPUcluster/home/UNCuser/relion2/relion/external/fltk/lib/libfltk.so
-- Found FFTW: fftw3
-- FFTW_LIBRARIES: /usr/lib64/libfftw3.so
-- Looking for sincos
-- Looking for sincos - found
-- Looking for __sincos
-- Looking for __sincos - not found
-- Building shared libs (smaller build size and binaries)
-- CMAKE_BINARY_DIR:/nas/GPUcluster/home/UNCuser/relion2/relion/build2
-- added autopick...
-- added autopick_mpi...
-- added display...
-- added find_tiltpairs...
-- added helix_toolbox...
-- added image_handler...
-- added maingui...
-- added manualpick...
-- added mask_create...
-- added particle_polish...
-- added particle_polish_mpi...
-- added particle_reposition...
-- added particle_sort...
-- added particle_sort_mpi...
-- added particle_symmetry_expand...
-- added pipeliner...
-- added postprocess...
-- added postprocess_mpi...
-- added preprocess...
-- added preprocess_mpi...
-- added project...
-- added reconstruct...
-- added refine...
-- added refine_mpi...
-- added run_ctffind...
-- added run_ctffind_mpi...
-- added run_motioncorr...
-- added run_motioncorr_mpi...
-- added stack_create...
-- added star_combine...
-- added star_compare...
-- added tiltpair_plot...
-- Configuring incomplete, errors occurred!
See also "/nas/GPUcluster/home/UNCuser/relion2/relion/build2/CMakeFiles/CMakeOutput.log".
See also "/nas/GPUcluster/home/UNCuser/relion2/relion/build2/CMakeFiles/CMakeError.log".

```

**Comment by bforsbe** [comment]
So you are trying to compile a relion binary on your computer (which has MPI), but then also on your cluster, which does not? 

The obvious discrepancy here is that MPI is entirely necessary for Relion, and if you have changed the cmake-files to try to omit MPI then I need to know what you changed. I guess what you would like to do is compile relion binaries and include MPI in them? In that case you need not change the cmake files, nor should you compile anything on the cluster. Is this what you want to do?

**Comment by UNCuser** [comment]
Sorry, I was unclear.  

> Of note, cluster has no mpi support, so open mpi was installed locally

What I should have said is open mpi was installed in my working directory on the cluster as I would like to compile relion2 on the cluster.

Thank you in advance.

**Comment by bforsbe** [comment]
Ok, so you want to specify a specific path for openmpi? Also did you change the CMakeLists.txt at all?

**Comment by UNCuser** [comment]
OK, looks like i got it all working.
Thank you so much for your help and patience.

FYI:
The path for mpicc and mpicxx was set working directory installation of open mpi.
Ah yes, I did change the CMakeList.txt -- added an include directories at line 163.
Ok so I took out line 163.
then 
> try commenting out line 169 from relion2-beta/CMakeLists.txt
> 
as previously described [https://github.com/3dem/relion/issues/55](url)

and was able to complete the cmake

Ran into one last snag - an additional issue as previously described [https://github.com/3dem/relion/issues/2](url)
So I just downloaded [FLTK](https://drive.google.com/uc?export=download&id=0B942d76zVnSeazZWcExRaXIyVDg) in my browser locally and then sftp/put the fltk into the externals/fltk* dir

So for so good.  Thanks so much.
cheers
